services/pgnotifier/pgnotifier.go

```diff
@@ -0,0 +1,447 @@
+		dbHandle: dbHandle,
+		URI:      connectionInfo,
+	}
+	err = notifier.setupQueue()
+	return
+}
+
+func (notifier PgNotifierT) AddTopic(topic string) (err error) {
+	stmt := fmt.Sprintf("DELETE FROM %s WHERE topic ='%s'", queueName, topic)
+	logger.Infof("PgNotifier: Deleting all jobs on topic: %s", topic)
+	_, err = notifier.dbHandle.Exec(stmt)
+	if err != nil {
+		return
+	}
+	err = notifier.createTrigger(topic)
+	if err != nil {
+		return
+	}
+	rruntime.Go(func() {
+		notifier.triggerPending(topic)
+	})
+	return
+}
+
+func (notifier *PgNotifierT) triggerPending(topic string) {
+	for {
+		time.Sleep(retriggerInterval)
+		stmt := fmt.Sprintf(`UPDATE %[1]s SET status='%[3]s',
+								updated_at = '%[2]s'
+								WHERE id IN  (
+								SELECT id
+								FROM %[1]s
+								WHERE status='%[3]s' OR status='%[4]s'
+								ORDER BY id
+								FOR UPDATE SKIP LOCKED
+								LIMIT %[5]v
+								);`,
+			queueName,
+			GetCurrentSQLTimestamp(),
+			WaitingState,
+			FailedState,
+			retriggerCount)
+		logger.Debugf("PgNotifier: triggering pending jobs")
+		_, err := notifier.dbHandle.Exec(stmt)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (notifier *PgNotifierT) trackBatch(batchID string, ch *chan []ResponseT) {
+	rruntime.Go(func() {
+		for {
+			time.Sleep(trackBatchInterval)
+			// keep polling db for batch status
+			// or subscribe to triggers
+			stmt := fmt.Sprintf(`SELECT count(*) FROM %s WHERE batch_id='%s' AND (status='%s' OR status='%s' OR status='%s')`, queueName, batchID, WaitingState, FailedState, ExecutingState)
+			var count int
+			err := notifier.dbHandle.QueryRow(stmt).Scan(&count)
+			if err != nil {
+				logger.Errorf("PgNotifier: Failed to query for tracking jobs by batch_id: %s, connInfo: %s", stmt, notifier.URI)
+				panic(err)
+			}
+			if count == 0 {
+				stmt = fmt.Sprintf(`SELECT payload, status FROM %s WHERE batch_id = '%s'`, queueName, batchID)
+				rows, err := notifier.dbHandle.Query(stmt)
+				if err != nil {
+					panic(err)
+				}
+				defer rows.Close()
+				responses := []ResponseT{}
+				for rows.Next() {
+					var payload json.RawMessage
+					var status string
+					err = rows.Scan(&payload, &status)
+					responses = append(responses, ResponseT{
+						Status:  status,
+						Payload: payload,
+					})
+				}
+				*ch <- responses
+				break
+			} else {
+				logger.Infof("PgNotifier: Pending %d files to process in batch: %s", count, batchID)
+			}
+		}
+	})
+}
+
+func (notifier *PgNotifierT) updateClaimedEvent(id int64, tx *sql.Tx, ch chan ClaimResponseT) {
+	rruntime.Go(func() {
+		response := <-ch
+		var err error
+		if response.Err != nil {
+			stmt := fmt.Sprintf(`UPDATE %[1]s SET status=(CASE
+					WHEN	attempt > %[2]d THEN CAST ( '%[3]s' AS pg_notifier_status_type)
+				ELSE  CAST( '%[4]s' AS pg_notifier_status_type)
+				END), updated_at = '%[5]s', error = '%[6]s' WHERE id = %[7]v`, queueName, maxAttempt, AbortedState, FailedState, GetCurrentSQLTimestamp(), response.Err.Error(), id)
```

Comment:
Instead of timestamp from the server, can we use the timestamp of the DB?

Reply:
we can user DB timestamp when we insert the row. To set it to DB timestamp for every update, we need to setup a postgres trigger and do in it